sdk/aqueduct/artifacts/utils.py

```diff
@@ -39,24 +44,53 @@ def preview_artifact(
     )
 
     preview_resp = api_client.__GLOBAL_API_CLIENT__.preview(dag=subgraph)
-    artifact_response = preview_resp.artifact_results[artifact_id]
 
-    serialization_type = artifact_response.serialization_type
+    assert target_artifact_id in preview_resp.artifact_results.keys()
+    target_artifact_result = preview_resp.artifact_results[target_artifact_id]
+    target_artifact_content = _get_content_from_artifact_result_resp(target_artifact_result)
+    target_artifact_type = target_artifact_result.artifact_type
+
+    existing_type_annotation = dag.must_get_artifact(target_artifact_id).type
+    if (
+        existing_type_annotation != ArtifactType.UNTYPED
+        and existing_type_annotation != target_artifact_type
+    ):
+        raise InvalidArtifactTypeException(
+            "The computed artifact is expected to be type %s, but has type %s"
+            % (existing_type_annotation, target_artifact_type)
+        )
+    dag.update_artifact_type(target_artifact_id, target_artifact_type)
+
+    # Any non-target artifacts are guaranteed to be upstream of the target artifact (due to the SubgraphDAGDelta),
+    # so if any of them are the result of a lazy operation, we'll want to backfill their types. *We do NOT backfill
+    # their contents*, as those are stored on the Artifact class itself and not the underlying shared dag.
+    for artifact_id, artifact_result in preview_resp.artifact_results.items():
+        # We've already processed the target artifact.
+        if artifact_id == target_artifact_id:
+            continue
+
+        dag.update_artifact_type(artifact_id, artifact_result.artifact_type)
```

Comment:
If an upstream artifact already has type set and there is a mismatch, I think we should also raise exception as in line 58?

Reply:
The only way this is possible if there is a preview cache miss on the upstream right? It's a little challenging to test this without an `ignore_cache` flag.

Reply:
@kenxu95 I think it might be possible to test if an artifact upstream involves a custom parameter?

Reply:
oh right I forgot parameter values are computed in the signature. Nice.